sshkey-attest/src/lib.rs

```diff
@@ -0,0 +1,453 @@
+//! This module contains the function that allows attestation of ssh sk keys which
+//! are created by FIDO2 devices.
+
+#![cfg_attr(docsrs, feature(doc_cfg))]
+#![cfg_attr(docsrs, feature(doc_auto_cfg))]
+#![deny(warnings)]
+#![warn(unused_extern_crates)]
+#![warn(missing_docs)]
+#![deny(clippy::todo)]
+#![deny(clippy::unimplemented)]
+#![deny(clippy::unwrap_used)]
+// #![deny(clippy::expect_used)]
+#![deny(clippy::panic)]
+#![deny(clippy::unreachable)]
+#![deny(clippy::await_holding_lock)]
+#![deny(clippy::needless_pass_by_value)]
+#![deny(clippy::trivially_copy_pass_by_ref)]
+
+#[macro_use]
+extern crate tracing;
+
+pub mod proto;
+
+use uuid::Uuid;
+use webauthn_rs_core::attestation::verify_attestation_ca_chain;
+use webauthn_rs_core::crypto::compute_sha256;
+use webauthn_rs_core::crypto::verify_signature;
+use webauthn_rs_core::internals::AuthenticatorData;
+use webauthn_rs_core::proto::COSEKey;
+use webauthn_rs_core::proto::ExtnState;
+use webauthn_rs_core::proto::RegisteredExtensions;
+use webauthn_rs_core::proto::Registration;
+use webauthn_rs_core::proto::{AttestationCaList, CredentialProtectionPolicy};
+
+use nom::bytes::complete::{tag, take};
+use nom::number::complete::be_u32;
+
+use openssl::{bn, ec, nid, x509};
+
+use webauthn_rs_core::error::WebauthnError;
+
+use webauthn_rs_core::proto::{
+    AttestationMetadata, COSEAlgorithm, COSEKeyType, ParsedAttestation, ParsedAttestationData,
+};
+
+use sshkeys::{Curve, EcdsaPublicKey, KeyType, KeyTypeKind, PublicKey, PublicKeyKind};
+
+use webauthn_rs_core::attestation::AttestationFormat;
+use webauthn_rs_core::attestation::{validate_extension, FidoGenCeAaguid};
+use webauthn_rs_core::crypto::assert_packed_attest_req;
+
+use crate::proto::AttestedPublicKey;
+
+/// Given an attestation generated by ssh-keygen, parse and validate it's content
+/// returning an attested public key structure that contains metadata and the ssh
+/// public key.
+///
+/// You can create an attested sk key with:
+/// ```shell
+///  dd if=/dev/urandom of=/Users/username/.ssh/id_ecdsa_sk.chal bs=16 count=1
+///  ssh-keygen -t ecdsa-sk -O challenge=/Users/username/.ssh/id_ecdsa_sk.chal \\
+///    -O write-attestation=/Users/username/.ssh/id_ecdsa_sk.attest -f /Users/username/.ssh/id_ecdsa_sk
+/// ```
+///
+/// The content of `id_ecdsa_sk.attest` and `id_ecdsa_sk.chal` correspond to the parameters
+/// `attestation` and `challenge` respectively.
+pub fn verify_fido_sk_ssh_attestation(
+    attestation: &[u8],
+    challenge: &[u8],
+    attestation_cas: &AttestationCaList,
+    danger_disable_certificate_time_checks: bool,
+) -> Result<AttestedPublicKey, WebauthnError> {
+    if attestation_cas.is_empty() {
+        return Err(WebauthnError::MissingAttestationCaList);
+    }
+
+    let alg = COSEAlgorithm::ES256;
+
+    // There doesn't seem to be much in the way of docs about the format of
```

Comment:
This comment seems to be about `ssh_sk_attestation_parser`; and should be in that function.

Also, it looks like there is documentation: https://github.com/openssh/openssh-portable/blob/2709809fd616a0991dc18e3a58dea10fb383c3f0/PROTOCOL.u2f#L151-L174